src/ImageSharp/Formats/Jpeg/JpegMetadata.cs

```diff
@@ -21,18 +27,94 @@ public JpegMetadata()
         /// <param name="other">The metadata to create an instance from.</param>
         private JpegMetadata(JpegMetadata other)
         {
-            this.Quality = other.Quality;
             this.ColorType = other.ColorType;
+
+            this.LuminanceQuantizationTable = other.LuminanceQuantizationTable;
+            this.ChromaQuantizationTable = other.ChromaQuantizationTable;
+            this.LuminanceQuality = other.LuminanceQuality;
+            this.ChrominanceQuality = other.ChrominanceQuality;
         }
 
         /// <summary>
-        /// Gets or sets the encoded quality.
+        /// Gets or sets luminance qunatization table for jpeg image.
         /// </summary>
-        public int Quality { get; set; } = 75;
+        internal Block8x8F LuminanceQuantizationTable
+        {
+            get
+            {
+                if (this.lumaQuantTable.HasValue)
+                {
+                    return this.lumaQuantTable.Value;
+                }
+
+                return Quantization.ScaleLuminanceTable(this.LuminanceQuality ?? Quantization.DefaultQualityFactor);
+            }
+
+            set => this.lumaQuantTable = value;
+        }
+
+        /// <summary>
+        /// Gets or sets chrominance qunatization table for jpeg image.
+        /// </summary>
+        internal Block8x8F ChromaQuantizationTable
+        {
+            get
+            {
+                if (this.chromaQuantTable.HasValue)
+                {
+                    return this.chromaQuantTable.Value;
+                }
+
+                return Quantization.ScaleChrominanceTable(this.ChrominanceQuality ?? Quantization.DefaultQualityFactor);
+            }
+
+            set => this.chromaQuantTable = value;
+        }
+
+        /// <summary>
+        /// Gets or sets the jpeg luminance quality.
+        /// </summary>
+        /// <remarks>
+        /// This value might not be accurate if it was calculated during jpeg decoding
+        /// with non-complient ITU quantization tables.
+        /// </remarks>
+        public int? LuminanceQuality { get; set; }
+
+        /// <summary>
+        /// Gets or sets the jpeg chrominance quality.
+        /// </summary>
+        /// <remarks>
+        /// This value might not be accurate if it was calculated during jpeg decoding
+        /// with non-complient ITU quantization tables.
+        /// </remarks>
+        public int? ChrominanceQuality { get; set; }
 
         /// <summary>
         /// Gets or sets the encoded quality.
         /// </summary>
+        /// <remarks>
+        /// Note that jpeg image can have different quality for luminance and chrominance components.
+        /// This property return average for both qualities and sets both qualities to the given value.
+        /// </remarks>
+        public int Quality
```

Comment:
Would be a breaking change, but wouldn't it be better to define this as `int?` to better communicate the `LuminanceQuality  == null && LuminanceQuality == null` case?

Reply:
Oops, forgot about it being `int?` in current state. Will work on it tomorrow with no public luma/chroma in mind.

Reply:
No it's `public int Quality` on the current master. I think we should consider changing it to `int?`, I doubt there are many users we would break.

Reply:
Yep. Break it. We cannot afford to hold onto properties that cause complexity. We simply don't have the maintainer numbers to manage continuously backwards compatible codebase. 

I also vote max or we never really fix anything due to loss of detail/extra noise.

Reply:
Got ya, thanks for your thoughts!

Reply:
While we are here breaking API: isn't it better to remove setter for metadata quality? There's no usage for it anywhere, jpegs are encoded using quality from `JpegEncoder`.

Reply:
Quality should be read from Meta and overwritten using the encoder however metadata is always writable for all formats. That means you should be able to set quality without spinning up a new encoder.

Reply:
Jpeg uses two qualities (luma/chroma) separately. `Quality` from metadata is just a public accessor for end user.

Reply:
I actually missed point for encoder, it's quality can be changed before each image encoding routine. Problem with metadata quality is that encoder first checks if its quality has value and only then takes quality from metadata (if it doesn't exists - take default value). So in a situation like this:

```csharp
var img = Image.Load(path);
img.Metadata.GetJpegMetadata().Quality = 100;

var encoder = new JpegEncoder { Quality = 80 };
img.SaveAsJpeg(path, encoder); // this would be saved quality=80 as encoder settings have higher priority
```

Metadata quality won't matter. It can affect encoding like this:

```csharp
var img = Image.Load(path);
img.Metadata.GetJpegMetadata().Quality = 100;

var encoder = new JpegEncoder {}; // we can even use as shared resource over multiple threads (when color stuff is fixed)
img.SaveAsJpeg(path, encoder); // this would be saved quality=100
```

Not sure if it was intended.

Reply:
Those examples you are providing are by design.

Upon review `Quality` in `JpegMetadata` should definitely not be nullable. It has a default value of `75` for when we create an image from scratch or change format. You can override that during decoding by taking the maximum of your calculated values.

The `Quality` property in the `JpegEncoder` is nullable and that's by design. This means we can always use the `JpegMetadata` property as a fallback but allow overriding it by choosing to use a new `JpegEncoder`. Choosing an explicit  `JpegEncoder` is a concious decision that falls outside of the normal workflow. Normally people will normally use the JpegEncoder in the `Configuration` instance.

Reply:
Can't agree with non-nullable quality in metadata. What's the usage of it for the user? `Quality` property is not used in the encoder with this PR. At the decoding stage separate luma/chroma qualities are stored (or their unique tables if those are not standard). At the encoding stage encoder tries to take quality from following sources in this exact order:

1. Encoder.Quality if is not null
2. Metadata.LuminanceQuality if is not null
3. Metadata.LuminanceQuantizationTable if is not null
4. DefaultQuality (75) <- this is a constant, it's not gathered from metadata

Same for Chrominance. We must save separate qualities because there are images with different qualities. `JpegMetadata.Quality` is simply a metadata for end user, it's not really used for encoding (only if user set `Quality` to some value which internally sets luma and chroma qualities to a given value). There's no uniform quality value anymore.

Reply:
>1. Encoder.Quality if is not null
>2. Metadata.LuminanceQuantizationTable if is not null
>3. Metadata.LuminanceQuality if is not null
>4. DefaultQuality (75) <- this is a constant in encoder code, it's not gathered from metadata


If you're doing it in that order what's the point of having the Quality property in the metadata at all?

I'm not convinced we should preserve the custom tables. Any manipulation of the incoming image will render any optimization performed that customization pointless.

Reply:
> If you're doing it in that order what's the point of having the Quality property in the metadata at all?

So we don't need to carry standard tables in the metadata, we can safely calculate them in the encoding routine using quality from decoder. And for actual metadata purposes if user wants it.

> I'm not convinced we should preserve the custom tables. Any manipulation of the incoming image will render any optimization performed that customization pointless.

Preserving custom tables is the least intrusive way of re-encoding the image. Does it affect encoded image? Yes. Does it affect it that much? No. Tested for 2x downscaling image. There is a difference, it ain't much but it does exist.

Any mid-priced camera tends to use custom tables for at least chrominance components so I don't see why we shouldn't preserve custom tables especially considering it is already done and is working correctly with existing tests. It doesnt really change anything internally except one method in the `JpegEncoder` and it doesn't affect performance. The only question left is `int Quality` or `int? Quality`.

Final word is yours so just yes/no on:
1. Custom tables
2. `int` or `int?`

And I will push final changes.

Reply:
> 1. Custom tables

We need to optimize to our most popular use case, which is thumbnail making. Assuming the user operation is a significant downscale, is the re-encoded thumbnail Jpeg "more correct" with the original tables or the default ones? I can't decide, such a downscale changes the frequency characteristics of the image a lot.

Sidenote: If I was to develop a thumbnail app, I would always set the Quality of the thumbnail to a higher value, even if it was low originally. Does ImageSharp.Web do anything like this @JimBobSquarePants?

> 2. `int` or `int?`

To me `metadata.Quality == null` would represent the case that the decoder was unable to estimate the quality of the input image nicely, while returning `metadata.Quality == 75` is quite an arbitrary choice. I vote for the breaking change (just let's make sure to apply the label).

Reply:
> We need to optimize to our most popular use case, which is thumbnail making. Assuming the user operation is a significant downscale, is the re-encoded thumbnail Jpeg "more correct" with the original tables or the default ones?

Barely noticable for 2x downsampling. So we have following situation:
Default tables:
**good** re-encoding with minimal change
**good** encoding of a heavily changed image

Original tables:
**even better** re-encoding with mininal change
**good** encoding of a heavily changed image

> To me `metadata.Quality == null` would represent the case that the decoder was unable to estimate the quality of the input image nicely, while returning `metadata.Quality == 75` is quite an arbitrary choice. I vote for the breaking change (just let's make sure to apply the label).

Not sure if we are talking about the same thing. `metadata.Quality` **is not used** during encoding process. Internal `metadata.LuminanceQuality` & `metadata.ChrominanceQuality` fields are used. Using `quality = max(luma, chroma)` would kill the quality if we are not talking about something extreme like thumbnails. Public quality is strictly a user way of knowing what was the decoded image quality. `metadata.Quality` setter sets luma & chroma to the same value. In other words, internal implementation won't really change and we either do (simplified):
`int quality = encoder.Quality ?? metadata.Quality ?? 75;` or `int quality = encoder.Quality ?? metadata.Quality;`

Reply:
@br3aker Ah sorry, my bad! I could have sworn I'd seen a change but I was wrong!

Thanks for doing the extra research. I agree with you on quality < 25. I really cannot see anyone doing that for anything other than experimental purpose.